### PR TITLE
Add Step Unit Testing Framework for Workflow Steps

### LINKED ISSUE #215: Step unit testing framework
## Description
Create a comprehensive unit testing framework specifically for workflow steps.

## Acceptance Criteria
- [ ] Test harness for individual steps
- [ ] Input/output validation helpers
- [ ] Performance testing support
- [ ] Coverage reporting for steps

Epic: epic/developer-tools

## Context
A dedicated testing framework would ensure workflow reliability through comprehensive testing.

### PR DESCRIPTION
**Pull request by parruda**
## Overview

This PR introduces a comprehensive unit testing framework specifically designed for testing individual workflow steps in Roast. This addresses the need for better testing infrastructure to ensure workflow reliability and step correctness.

## Why This Feature is Needed

Currently, testing workflow steps requires running entire workflows, making it difficult to:
- Isolate and test individual step behavior
- Debug step failures efficiently  
- Validate step inputs and outputs
- Monitor step performance
- Track test coverage for step implementations

### 🧪 Test Harness (`StepTestHarness`)
- Provides isolated execution environment for individual steps
- Supports mocking of dependencies and external services
- Enables step testing without full workflow context

### ✅ Input/Output Validators (`StepValidators`)
- Schema validation for step inputs and outputs
- Type checking and constraint validation
- Custom validation rule support

### 📊 Performance Monitoring (`PerformanceMonitor`)
- Execution time tracking and analysis
- Memory usage monitoring
- Performance regression detection
- Benchmarking capabilities

### 📈 Coverage Reporting (`StepCoverage`)
- Tracks which steps have been tested
- Reports coverage metrics for step implementations
- Integrates with existing testing infrastructure

### 🛠️ Test Case Framework (`StepTestCase`)
- Extends Minitest with step-specific assertions
- Provides convenient setup/teardown for step testing
- Includes common test patterns and helpers

### Basic Step Testing
```ruby
class MyStepTest < Roast::Workflow::Testing::StepTestCase
  def test_step_execution
    step = create_step("my_step", input: { value: 42 })
    result = execute_step(step)
    
    assert_step_success(result)
    assert_output_matches(result, { processed_value: 84 })
  end
end
```

### Performance Testing
```ruby
def test_step_performance
  monitor = performance_monitor
  
  step = create_step("expensive_step")
  result = monitor.measure { execute_step(step) }
  
  assert_execution_time_under(result, 5.seconds)
  assert_memory_usage_under(result, 100.megabytes)
end
```

### Input Validation Testing
```ruby
def test_input_validation
  validator = step_validator_for("data_processor")
  
  assert_valid_input(validator, { data: [1, 2, 3] })
  assert_invalid_input(validator, { data: "invalid" })
end
```

## Benefits for Developers

- **Faster Development**: Test individual steps without running full workflows
- **Better Debugging**: Isolate issues to specific step implementations
- **Improved Reliability**: Comprehensive validation of step behavior
- **Performance Insights**: Monitor and optimize step performance
- **Higher Confidence**: Detailed coverage reporting for step testing

## Implementation Details

- Built on top of existing Minitest infrastructure
- Follows Roast's architectural patterns and conventions
- Integrates with current workflow execution system
- Maintains backward compatibility with existing tests
- Includes comprehensive documentation and examples

## Testing

The framework includes extensive tests covering:
- All core testing components
- Integration with workflow system
- Error handling and edge cases
- Performance monitoring accuracy
- Coverage reporting functionality

## Documentation

- Complete API documentation for all components
- Usage examples in `examples/step_testing/`
- Integration guides and best practices
- Migration guide for existing test suites

Closes #215

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: lib/roast/workflow/testing.rb [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+          end
+
+          # Check required methods
+          step_instance = begin
+            step_class.new(MockWorkflow.new)
+          rescue
+            nil
+          end
+          if step_instance
+            [:call, :name, :workflow].each do |method|
```
Comment (ericproulx) [inline lib/roast/workflow/testing.rb:150]: I've seen this a lot in the code base and it should be a constant instead. Maybe giving claude more precise instruction about that and it won't repeat it. I'll give it a try.

File: lib/roast/workflow/testing/step_test_case.rb [hunk 2]
```diff
@@ -0,0 +1,275 @@
+# frozen_string_literal: true
+
+require "active_support/test_case"
```
Comment (ericproulx) [inline lib/roast/workflow/testing/step_test_case.rb:3]: No need for that

**Review by juniper-shopify — CHANGES_REQUESTED** [review]
Given the massive changes/refactors we're introducing with the new Roast DSL, I recommend we don't go down the path of making this change to the legacy version.
